File: 322-coin-change/322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount+1];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;
        for(int i=1; i<=amount; i++) {
            for(int j:coins) {
                if(j <= i && dp[i-j] != Integer.MAX_VALUE) {
                    dp[i] = Math.min(dp[i], 1 + dp[i-j]);
                }
            }
        }
        return dp[amount] == Integer.MAX_VALUE ? -1 : dp[amount];
    }
}

File: 322-coin-change/322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount+1] * (amount+1)
        dp[0] = 0
        
        for coin in coins:    
            for cash in range(amount+1):
                if(coin <= cash):
                    dp[cash] = min(dp[cash], 1 + dp[cash - coin])
        
        return -1 if dp[amount] > amount else dp[amount]